File: juno/indicators/smma.py
from decimal import Decimal

from .sma import Sma


# Smoothed Moving Average
class Smma:
    value: Decimal = Decimal('0.0')
    _sma: Sma
    _weight: int
    _t: int = 0
    _t1: int
    _t2: int

    def __init__(self, period: int) -> None:
        self._sma = Sma(period)
        self._weight = period
        self._t1 = period
        self._t2 = period + 1

    @property
    def maturity(self) -> int:
        return self._t1

    @property
    def mature(self) -> bool:
        return self._t >= self._t1

    def update(self, price: Decimal) -> Decimal:
        self._t = min(self._t + 1, self._t2)

        if self._t <= self._t1:
            self._sma.update(price)

        if self._t == self._t1:
            self.value = self._sma.value
        elif self._t >= self._t2:
            self.value = (self.value * (self._weight - 1) + price) / self._weight

        return self.value

File: juno_rs/src/indicators/smma.rs
use super::{
    sma::{Sma, SmaParams},
    MA,
};
use serde::{Deserialize, Serialize};
use std::cmp::min;

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct SmmaParams {
    pub period: u32,
}

pub struct Smma {
    pub value: f64,

    sma: Sma,
    weight: f64,

    t: u32,
    t1: u32,
    t2: u32,
}

impl Smma {
    pub fn new(params: &SmmaParams) -> Self {
        // Period validated within Sma.
        Self {
            value: 0.0,
            sma: Sma::new(&SmaParams {
                period: params.period,
            }),
            weight: params.period.into(),
            t: 0,
            t1: params.period,
            t2: params.period + 1,
        }
    }
}

impl MA for Smma {
    fn maturity(&self) -> u32 {
        self.t2
    }

    fn mature(&self) -> bool {
        self.t >= self.t2
    }

    fn update(&mut self, price: f64) {
        self.t = min(self.t + 1, self.t2);

        if self.t <= self.t1 {
            self.sma.update(price);
        }

        if self.t == self.t1 {
            self.value = self.sma.value;
        } else if self.t >= self.t2 {
            self.value = (self.value * (self.weight - 1.0) + price) / self.weight;
        }
    }

    fn value(&self) -> f64 {
        self.value
    }
}
